src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,479 @@
+ *   encounter it locally if needed. Note that we may cast the var to some certain dtype multiple
+ *   times, but we decide not to store and reuse the casted copy due to the stroage concern and to
+ *   be more friendly to inlining and operator fusion.
+ *
+ * The information of each op is registered in the Op::GetAttr<FMixedPrecision>("FMixedPrecision").
+ * The registered function has signature: FMixedPrecision. We will call the registered function
+ * with the original call and the global output_dtype parameter. The registered function will
+ * return the policy of the op, whether the op can adjust the dtype of the accumulator, and the
+ * new call node with output_dtype set to the global output_dtype parameter.
+ *
+ * Key design: wrap_param op
+ *   We need to use fp16 parameters (which appear as constants in the program), but the type
+ *   inference will fail if some parameters are fp16 and some are fp32 in the original module. To
+ *   solve this, we introduce a new op wrap_param, which will wrap the original parameter and cast
+ *   it to fp32 var.
+ *
+ *   When we encounter the var afterwards, we will directly replace it with the parameter. This
+ *   information is tracked by the const_map_.
+ */
+class DTypeDecisionCollector : public ExprVisitor {
+ public:
+  explicit DTypeDecisionCollector(DataType output_dtype) : output_dtype_(output_dtype) {}
+
+  static VarDTypeMap Collect(Function func, DataType output_dtype) {
+    DTypeDecisionCollector collector(output_dtype);
+    collector.VisitExpr(func);
+    return std::move(collector.var_dtype_map_);
+  }
+
+ private:
+  // merge the message for a var
+  void UpdateVarDTypeMap(const Var& var, const NDType& dtype) {
+    auto it = var_dtype_map_.find(var);
+    if (it == var_dtype_map_.end()) {
+      var_dtype_map_[var] = dtype;
+    } else {
+      var_dtype_map_[var] = NDTypeMerge(it->second, dtype);
+    }
+  }
+
+  // merge the message for all vars in the expr list
+  void RequireArgsToType(Array<Expr> args, Array<NDType> to) {
+    ICHECK(args.size() == to.size()) << "Invalid target dtypes";
+    for (size_t i = 0; i < args.size(); ++i) {
+      auto fvisitleaf = [&](const Expr& expr, NDType to) {
+        if (const auto* var = expr.as<VarNode>()) {
+          UpdateVarDTypeMap(GetRef<Var>(var), to);
+        } else if (expr->IsInstance<ConstantNode>()) {
+          // Constant can be casted anyway, so we don't need to do anything here
+          return;
+        } else {
+          LOG(FATAL) << "Unsupported argument type: " << expr->GetTypeKey();
+        }
+      };
+      DecomposeNestedMsg(args[i], to[i], fvisitleaf);
+    }
+  }
+
+  // merge the message for all vars in the expr list
+  void RequireArgsToType(Array<Expr> args, DataType to) {
+    std::vector<NDType> to_arr;
+    for (const Expr& arg : args) {
+      // Note that this implicitly requires all arg to be nested tensors.
+      to_arr.push_back(NDTypeFrom(arg, to));
```

Comment:
What if some argument is not Tensor or Tuple of Tensors? Like `reshape`, `full`, `broadcast_to`, etc. `NDTypeFrom` now requires every leaf to be Tensor so seems that we should skip the non-Tensor arguments in some way?